### ISSUE TITLE
Parse EPA and EIA data into one table of generator data to be used in the HIFLD Grid

### ISSUE THREAD

**Issue by danielolsen** [Breakthrough-Energy/PreREISE#222]
# :rocket:

- [x] Is your feature request essential for your project?

### Describe the workflow you want to enable
I wish I could parse data from the EIA and EPA, and align them, to be able to estimate all relevant information for generators (units) or power plants (collections of units).

### Describe your proposed implementation

The EPA's AMPD dataset has hourly samples of (power, heat) for units, which can be used to fit a heat rate curve. The EPA's NEEDS dataset has linear heat-rate estimates as a fall-back. See the EPA Power Sector Modeling Platform v6 documentation for the [November 2018](https://www.epa.gov/airmarkets/documentation-epas-power-sector-modeling-platform-v6-november-2018-reference-case) and [January 2020](https://www.epa.gov/airmarkets/documentation-epas-power-sector-modeling-platform-v6-january-2020-reference-case) versions.

The EIA data has other information we want, like minimum generation level and more details on generator types (e.g. standalone CT vs. standalone ST vs. units that are part of a combined-cycle plant).

The data are not well aligned between EIA and EPA data sets, because the EIA cares about what can make electricity (generators) while the EPA cares about what can make emissions (boilers and CTs), and they both refer to them as 'units', even though the mapping is not always 1-to-1. There is a project that aims to aid in mapping: https://github.com/USEPA/camd-eia-crosswalk, if we can figure out how to use the information effectively.

There is also an open question of at which level we want to be aggregating units, especially for combined-cycle plants. As an example, the West Phoenix plant in Arizona has 3x CT/ST pairs: 1x (91 MW, 45 MW) + 2x (185 MW, 201 MW). Do we want to try to create: a) one plant, containing all generators, b) two plants, one for the smaller CC pair and one for the combination of the identical larger two, c) three plants, one for each CC pair, using each pair's fitted heat rate curve, d) three plants, one for the smaller CC pair, and two identical ones for the larger CC pairs, where we use data from both to fit a single heat rate curve, e) six individual plants, f) something else?

a) Benefits: simple. Drawbacks: overly simplified, different CC units have different characteristics.
b) Benefits: better resolution than (a). Drawbacks: we have to assume that the two larger CCs are identical based on their nameplate capacities (we could also look at things like operational date, but would have to make some assumptions about how close is close enough).
c) Benefits: better resolution than (b). Drawbacks: less data to fit curves for each of the two larger generators
d) Benefits: likely better data quality than (c). Drawbacks: like (b), we need to make assumptions about which generators are similar enough to be considered identical
e) Benefits: better resolution. Drawbacks: loss of connection between units that are connected in reality, less data to fit heat rate curves.

At a higher level, how do we want to parse through the ~20k generators to decide which units to aggregate?

Hopefully this Issue can serve as the hub for brainstorming approaches and discussing challenges in implementation.

**Comment by danielolsen** [comment]
There's a lot of overlapping information between the HIFLD datasets (Power_Plants.csv, Generating_Units.csv) and the EIA Form 860 data (Sheet `2___PlantYyyy` for plants, Sheet `3_1_GeneratorYyyyy` for units). Looking through my current code for #197, it seems that so far we're looking at the HIFLD data columns:
- Power_Plants.csv
  - status
  - state
  - county
  - latitude
  - longitude
  - PLANT_CODE
- Generating_Units.csv
  - status
  - state
  - ZIP
  - type
  - nameplate capacity
  - PLANT_CODE

Then it seems we only use the EIA data to look up Pmin values. I don't see power plant status explicitly as a part of the EIA Form 860 data, but for everything else there appears to be equivalent information in the EIA data. I would generally have more trust in the EIA data than HIFLD for generators, so I wonder if it makes sense to use all EIA/EPA data for coming up with our generator data set, and only using the HIFLD data for the transmission network. I think it would be simple enough to accomplish: all of our logic and data structures stay the same, we just need to load a different table to start with, and rename a few columns.

**Comment by danielolsen** [comment]
If we start from the EIA Form 860 units, we can map 87% of the total summer capacity to at least _some_ data in the EPA AMPD dataset (haven't looked at how much yet or whether it's good enough to generate a heat rate curve from):
```python
>>> import pandas as pd
>>> from prereise.gather.griddata.hifld import const
>>> from prereise.gather.griddata.hifld.data_access import load
>>>
>>> # Helper function
>>> def floatify(x):
...     try:
...         return float(x.replace(",", ""))
...     except ValueError:
...         if x == " ":
...             return pd.NA
...         else:
...             print(repr(x))
...             return pd.NA
...
>>> # Constants
>>> no_ampd_expected_techs = {
...     "Batteries",
...     "Conventional Hydroelectric",
...     "Flywheels",
...     "Geothermal",
...     "Hydroelectric Pumped Storage",
...     "Nuclear",
...     "Offshore Wind Turbine",
...     "Onshore Wind Turbine",
...     "Solar Photovoltaic",
...     "Solar Thermal with Energy Storage",
...     "Solar Thermal without Energy Storage",
... }
>>>
>>> # Load raw tables
>>> # This one takes a while
>>> epa_ampd = load.get_epa_ampd(const.blob_paths["epa_ampd"])
sys:1: DtypeWarning: Columns (10,12,18,20) have mixed types.Specify dtype option on import or set low_memory=False.
sys:1: DtypeWarning: Columns (10,12,14,16,18,20) have mixed types.Specify dtype option on import or set low_memory=False.
sys:1: DtypeWarning: Columns (3,10,12,18,20) have mixed types.Specify dtype option on import or set low_memory=False.
>>> form_860 = load.get_eia_form_860(const.blob_paths["eia_form860_2019"])
sys:1: DtypeWarning: Columns (15,37,38,42) have mixed types.Specify dtype option on import or set low_memory=False.
>>> crosswalk = pd.read_csv("https://github.com/USEPA/camd-eia-crosswalk/releases/download/v0.2.1/epa_eia_crosswalk.csv")
>>>
>>> # Initial processing of the EPA AMPD
>>> epa_ampd = epa_ampd.astype({"UNITID": "string"})
>>> epa_units = set(epa_ampd.groupby(["ORISPL_CODE", "UNITID"]).groups.keys())
>>> print(len(epa_units))
4313
>>>
>>> # Initial processing of Form 860
>>> form_860 = form_860.astype({"Plant Code": int})
>>> form_860["Summer Capacity (MW)"] = form_860["Summer Capacity (MW)"].map(floatify)
>>> print(len(form_860))
21971
>>>
>>> # Initial processing of EIA/EPA Crosswalk
>>> crosswalk_match_exclude = {"CAMD Unmatched", "Manual CAMD Excluded"}
>>> crosswalk = crosswalk.query(
...     "MATCH_TYPE_GEN not in @crosswalk_match_exclude and CAMD_STATUS != 'RET'"
... ).astype(
...     {
...         "MOD_EIA_PLANT_ID": int,
...         "MOD_CAMD_UNIT_ID": "string",
...         "MOD_CAMD_GENERATOR_ID": "string",
...         "MOD_EIA_GENERATOR_ID_GEN": "string",
...     }
... )
>>> print(len(crosswalk))
5305
>>>
>>> # Build translation Series
>>> crosswalk_translation = crosswalk.set_index(
...     ["MOD_EIA_PLANT_ID", "MOD_CAMD_GENERATOR_ID"]
... ).apply(
...     lambda x: (x["CAMD_PLANT_ID"], x["MOD_CAMD_UNIT_ID"]),
...     axis=1,
... )
>>> print(crosswalk_translation)
MOD_EIA_PLANT_ID  MOD_CAMD_GENERATOR_ID
3                 1                               (3, 1)
                  2                               (3, 2)
                  4                               (3, 4)
                  5                               (3, 5)
                  A1CT                           (3, 6A)
                                               ...
60903             3                           (60903, 1)
                  2                           (60903, 2)
                  4                           (60903, 2)
70454             MAG1                     (70454, MAG1)
                  MAG2                     (70454, MAG2)
Length: 5305, dtype: object
>>>
>>> # Filter translation to just what's in the EPA data
>>> crosswalk_with_ampd_data = crosswalk_translation.loc[crosswalk_translation.isin(epa_units)]
>>> print(crosswalk_with_ampd_data)
MOD_EIA_PLANT_ID  MOD_CAMD_GENERATOR_ID
3                 1                               (3, 1)
                  2                               (3, 2)
                  4                               (3, 4)
                  5                               (3, 5)
                  A1CT                           (3, 6A)
                                               ...
60903             3                           (60903, 1)
                  2                           (60903, 2)
                  4                           (60903, 2)
70454             MAG1                     (70454, MAG1)
                  MAG2                     (70454, MAG2)
Length: 5196, dtype: object
>>>
>>> # Calculate how much of the Form 860 capacity maps to EPA data via crosswalk
>>> form_860_keys = form_860[["Plant Code", "Generator ID"]].apply(tuple, axis=1)
>>> print(
...     "EIA Capacity mapped to CAMD (MW):",
...     "{:,.0f}".format(
...         form_860.loc[form_860_keys.isin(crosswalk_with_ampd_data.index), "Summer Capacity (MW)"].sum()
...     ),
... )
EIA Capacity mapped to CAMD (MW): 651,590
>>> print(
...     "EIA Capacity expected to be mapped to CAMD (MW):",
...     "{:,.0f}".format(
...         form_860.loc[~form_860["Technology"].isin(no_ampd_expected_techs), "Summer Capacity (MW)"].sum()
...     ),
... )
EIA Capacity expected to be mapped to CAMD (MW): 749,216
```

For the generators without data (97.6 GW), there's approximately:
- 25 GW open-loop gas-fired combustion turbines (19% of total capacity for this type)
- 15 GW gas-fired combustion turbines as part of a combined-cycle plant (10% of total)
- 10 GW steam coal (4.4% of total)
- 9.8 GW gas-fired steam turbine as part of a combined-cycle plant (9.5% of total)
- 6.6 GW wood or wood-waste biomass steam turbine (**79% of total**)
- 5.4 GW gas-fired steam turbine (standalone) (7.7% of total)
- 5.1 GW petroleum-liquids-fired gas turbine (**38% of total**)
- 4.8 GW gas-fired internal-combustion (**100% of total**)
- 4.6 GW petroleum-liquids-fired internal-combustion (**100% of total**)
- 2.1 GW gas-fired single-shaft combined-cycle (16% of total)
- 2.0 GW municipal solid waste steam turbines (**98% of total**)
- 1.4 GW landfill gas internal combustion (**100% of total**)
- 5.3 GW other

Assuming we can get good heat rate curves for a majority of the plants with CAMD data, I think we can fill in the missing data reasonably well using the data for that type, e.g. with weighted-average heat-rates or with capacity-to-heat-rate regressions. The classes of generators with mostly missing data seem to fall into either biomass, municipal solid waste, landfill gas, or rarely used backup generators (my assumption for petroleum-liquids-fired gas turbine, gas-fired internal-combustion, and petroleum-liquids-fired internal-combustion). We can probably come up with some decent assumptions for these in another way, and they're probably not especially important to dispatch or capacity decisions.

**Comment by danielolsen** [comment]
One more piece of information that we'll need to transform heat rate curves to cost curves: fuel prices. One approach would be to add per-zone fuel prices to **const.py**.

**Comment by danielolsen** [comment]
Closed via #225. Extra features may be added later.
